### PR TITLE
Added support for ObjectIds in fields other than _id & improved date casting

### PR DESCRIPTION
**Pull request by Flambe**
I've extended the $casts property so that it's now possible to specify fields as 'objectid'. If you specify a field as 'objectid', it will be converted to one when you save/search for something.

Example syntax:
```php
class User extends .../Model {
    protected $casts = [
        'book_id' => 'objectid'
    ];
}
````

This PR also fixes issues with the $dates array not being followed when using dot notation. 

This should solve #356 as you can now specify relationships in the $casts array. I decided not to make it the default behaviour for relationships as Mongo does not enforce relationship types, and as to not break backwards compatibility.

(other related issues: #134 #449 #350)

### CODE REVIEW

**Comment by keanass** [comment]
why Some checks were not successful  ? please we need this feature asap can you check it +1

**Comment by Djuki** [comment]
@jenssegers @Flambe  What are the obstacles for this PR to be merged? Is there something we can do to help testing or additional work?

**Comment by Flambe** [comment]
> @jenssegers @Flambe What are the obstacles for this PR to be merged? Is there something we can do to help testing or additional work?

I'm pretty sure there are a few edge cases with the casting that I've missed. If you find any, I'll try and get them fixed, or you could make a PR on the repo for this PR. Otherwise I'm not sure what needs to be done.

**Comment by bytewoopy** [comment]
Did anyone manage to solve this?

**Comment by Flambe** [comment]
@bytewoopy We're still using this branch for our own usage of the library. We haven't been putting in an active effort to keep it ready to merge as we've had no feedback from the maintainers.

Besides a few edge cases with casting, it's working as intended.

**Comment by Smolevich** [comment]
@Flambe do you want merge this PR into stable release?

**Comment by Flambe** [comment]
@Smolevich I'd still like to get it merged, though it needs updating first and I haven't got around to that yet.

**Comment by Smolevich** [comment]
> @Smolevich I'd still like to get it merged, though it needs updating first and I haven't got around to that yet.

Ok, ping if you resolve conflicts

**Comment by Flambe** [comment]
Will do, though it will likely be after the new year

**Comment by divine** [comment]
@Flambe laravel 7 has merged custom cast types ( see https://github.com/laravel/framework/pull/31035 ), will it help without modifying much of code?

@Smolevich what do you think?

Thank you!

**Comment by Smolevich** [comment]
Looks like as it can help, but needs to investigate

**Comment by Flambe** [comment]
I completely forgot about the conflicts, I'll try and get them done soon if I have some time!

As Smolevich said, it'll need investigation if that'll help. Using that would also restrict us to Laravel 7 I believe.

**Comment by divine** [comment]
> I completely forgot about the conflicts, I'll try and get them done soon if I have some time!

I myself will be waiting for this too so fingers crossed 🙏 .

Thank you for your contribution!

**Comment by Jannnnnn** [comment]
A improvement idea, use a const instead of `someId => 'objectid'` `someId => self::CAST_OBJECTID`

File: src/Jenssegers/Mongodb/Eloquent/Model.php [hunk 1]
```diff
@@ -480,4 +686,39 @@ public function __call($method, $parameters)
 
         return parent::__call($method, $parameters);
     }
+
+    /**
+     * @inheritdoc
```
Comment (Smolevich) [inline src/Jenssegers/Mongodb/Eloquent/Model.php]: Fill params please in dock block

File: src/Jenssegers/Mongodb/Eloquent/Model.php [hunk 2]
```diff
@@ -338,11 +461,94 @@ protected function pushAttributeValues($column, array $values, $unique = false)
...
+            } else {
+                $attributes[$key] = $this->serializeDate($this->asDateTime($attributes[$key]));
+            }
+        }
+
+        return $attributes;
+    }
+
+    /**
+     * @inheritdoc
```
Comment (Smolevich) [inline src/Jenssegers/Mongodb/Eloquent/Model.php]: Add dock block for `$key`

**Comment by rennokki** [comment]
That's a freaking huge PR. I think it has to wait until a further stage of the package.

**Comment by divine** [comment]
Hi there,

This feature will be released within next version.

Closing for now.

Thanks.

**Comment by Flambe** [comment]
Thank you @divine! Sorry for not getting to this sooner.

**Pull request closed without merging**